app/Services/ProvisioningService.php

```diff
@@ -0,0 +1,366 @@
+        'server_pool' => 'required|string|exists:server_pools,name',
+    ];
+
+    public function create(object $properties)
+    {
+        $this->createWrapper($properties, function ($type) use ($properties) {
+            $type->name = $properties->name;
+            $type->description = $properties->description;
+            $type->color = $properties->color;
+            $pool = ServerPool::firstWhere('name', $properties->server_pool);
+            $type->serverPool()->associate($pool);
+        });
+    }
+
+    public function destroy(array $match = [])
+    {
+        $this->destroyWrapper($match);
+    }
+}
+
+class RoleProvisioner extends AbstractProvisioner
+{
+    protected string $model = Role::class;
+
+    protected array $expectedProperties = [
+        'name' => 'required|string',
+        'permissions' => 'required|array:rooms,meetings,settings,users,roles,roomTypes,servers,serverPools',
+        'permissions.rooms' => 'list',
+        'permissions.rooms.*' => 'string',
+        'permissions.meetings' => 'list',
+        'permissions.meetings.*' => 'string',
+        'permissions.settings' => 'list',
+        'permissions.settings.*' => 'string',
+        'permissions.users' => 'list',
+        'permissions.users.*' => 'string',
+        'permissions.roles' => 'list',
+        'permissions.roles.*' => 'string',
+        'permissions.roomTypes' => 'list',
+        'permissions.roomTypes.*' => 'string',
+        'permissions.servers' => 'list',
+        'permissions.servers.*' => 'string',
+        'permissions.serverPools' => 'list',
+        'permissions.serverPools.*' => 'string',
+    ];
+
+    public function create(object $properties)
+    {
+        $this->createWrapper($properties, function ($role) use ($properties) {
+            foreach ($properties->permissions as $group => $perms) {
+                foreach ($perms as $item) {
+                    $permName = "$group.$item";
+                    $perm = Permission::firstWhere('name', $permName);
+                    if (is_null($perm)) {
+                        throw new RecordsNotFoundException("Could not find permission with name '$permName'");
+                    }
+                    $permissions[] = $perm->id;
+                }
+            }
+            $role->name = $properties->name;
+            $role->save();
+            $role->permissions()->sync($permissions);
+        });
+    }
+
+    public function destroy(array $match = [])
+    {
+        $this->destroyWrapper($match);
+    }
+}
+
+class UserProvisioner extends AbstractProvisioner
+{
+    protected string $model = User::class;
+
+    protected array $expectedProperties = [
+        'firstname' => 'required|string',
+        'lastname' => 'required|string',
+        'email' => 'required|string',
+        'password' => 'required|string',
+        'authenticator' => 'required|string',
+        'roles' => 'required|list|exists:roles,name',
+        'roles.*' => 'string',
```

Comment:
```suggestion
        'roles' => 'required|list',
        'roles.*' => 'string|distinct|exists:roles,name',
```
See my comment in ServerPoolProvisioner

Reply:
Fixed.